src/core/document.js

```diff
@@ -280,10 +287,23 @@ var Page = (function PageClosure() {
           return pageOpList;
         }
 
-        var annotationsReadyPromise = Annotation.appendToOperatorList(
-          annotations, pageOpList, partialEvaluator, task, intent,
-          renderInteractiveForms);
-        return annotationsReadyPromise.then(function () {
+        // Collect the operator list promises for the annotations. Each promise
+        // is resolved with the complete operator list for a single annotation.
+        var i, ii;
+        var opListPromises = [];
+        for (i = 0, ii = annotations.length; i < ii; i++) {
+          if (isAnnotationRenderable(annotations[i], intent)) {
+            opListPromises.push(annotations[i].getOperatorList(
+              partialEvaluator, task, renderInteractiveForms));
+          }
+        }
+
+        return Promise.all(opListPromises).then(function(opLists) {
+          pageOpList.addOp(OPS.beginAnnotations, []);
+          for (i = 0, ii = opLists.length; i < ii; i++) {
+            pageOpList.addOpList(opLists[i]);
+          }
+          pageOpList.addOp(OPS.endAnnotations, []);
```

Comment:
Nit: For improved readability, considering the importance of `flush`ing the operatorList, I think that it'd be good to add a new line after the annotation OPS here.